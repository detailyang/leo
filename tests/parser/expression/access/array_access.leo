/*
namespace: ParseExpression
expectation: Pass
*/

x[0]
X[1]
x[0u8]
x[1u8][2u8]
x[x][y][z]
x[0]()
x()[0]
x(y)::y(x)
x[x].0[x]
