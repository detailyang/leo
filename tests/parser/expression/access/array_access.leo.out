---
namespace: ParseExpression
expectation: Pass
outputs:
  - ArrayAccess:
      array:
        Identifier: "{\"name\":\"x\",\"span\":\"{\\\"line_start\\\":1,\\\"line_stop\\\":1,\\\"col_start\\\":1,\\\"col_stop\\\":2,\\\"path\\\":\\\"array_access.leo\\\",\\\"content\\\":\\\"x[0]\\\"}\"}"
      index:
        Value:
          Implicit:
            - "0"
            - line_start: 1
              line_stop: 1
              col_start: 3
              col_stop: 4
              path: array_access.leo
              content: "x[0]"
      span:
        line_start: 1
        line_stop: 1
        col_start: 1
        col_stop: 5
        path: array_access.leo
        content: "x[0]"
  - ArrayAccess:
      array:
        Identifier: "{\"name\":\"X\",\"span\":\"{\\\"line_start\\\":1,\\\"line_stop\\\":1,\\\"col_start\\\":1,\\\"col_stop\\\":2,\\\"path\\\":\\\"array_access.leo\\\",\\\"content\\\":\\\"X[1]\\\"}\"}"
      index:
        Value:
          Implicit:
            - "1"
            - line_start: 1
              line_stop: 1
              col_start: 3
              col_stop: 4
              path: array_access.leo
              content: "X[1]"
      span:
        line_start: 1
        line_stop: 1
        col_start: 1
        col_stop: 5
        path: array_access.leo
        content: "X[1]"
  - ArrayAccess:
      array:
        Identifier: "{\"name\":\"x\",\"span\":\"{\\\"line_start\\\":1,\\\"line_stop\\\":1,\\\"col_start\\\":1,\\\"col_stop\\\":2,\\\"path\\\":\\\"array_access.leo\\\",\\\"content\\\":\\\"x[0u8]\\\"}\"}"
      index:
        Value:
          Integer:
            - U8
            - "0"
            - line_start: 1
              line_stop: 1
              col_start: 3
              col_stop: 6
              path: array_access.leo
              content: "x[0u8]"
      span:
        line_start: 1
        line_stop: 1
        col_start: 1
        col_stop: 7
        path: array_access.leo
        content: "x[0u8]"
  - ArrayAccess:
      array:
        ArrayAccess:
          array:
            Identifier: "{\"name\":\"x\",\"span\":\"{\\\"line_start\\\":1,\\\"line_stop\\\":1,\\\"col_start\\\":1,\\\"col_stop\\\":2,\\\"path\\\":\\\"array_access.leo\\\",\\\"content\\\":\\\"x[1u8][2u8]\\\"}\"}"
          index:
            Value:
              Integer:
                - U8
                - "1"
                - line_start: 1
                  line_stop: 1
                  col_start: 3
                  col_stop: 6
                  path: array_access.leo
                  content: "x[1u8][2u8]"
          span:
            line_start: 1
            line_stop: 1
            col_start: 1
            col_stop: 7
            path: array_access.leo
            content: "x[1u8][2u8]"
      index:
        Value:
          Integer:
            - U8
            - "2"
            - line_start: 1
              line_stop: 1
              col_start: 8
              col_stop: 11
              path: array_access.leo
              content: "x[1u8][2u8]"
      span:
        line_start: 1
        line_stop: 1
        col_start: 1
        col_stop: 12
        path: array_access.leo
        content: "x[1u8][2u8]"
  - ArrayAccess:
      array:
        ArrayAccess:
          array:
            ArrayAccess:
              array:
                Identifier: "{\"name\":\"x\",\"span\":\"{\\\"line_start\\\":1,\\\"line_stop\\\":1,\\\"col_start\\\":1,\\\"col_stop\\\":2,\\\"path\\\":\\\"array_access.leo\\\",\\\"content\\\":\\\"x[x][y][z]\\\"}\"}"
              index:
                Identifier: "{\"name\":\"x\",\"span\":\"{\\\"line_start\\\":1,\\\"line_stop\\\":1,\\\"col_start\\\":3,\\\"col_stop\\\":4,\\\"path\\\":\\\"array_access.leo\\\",\\\"content\\\":\\\"x[x][y][z]\\\"}\"}"
              span:
                line_start: 1
                line_stop: 1
                col_start: 1
                col_stop: 5
                path: array_access.leo
                content: "x[x][y][z]"
          index:
            Identifier: "{\"name\":\"y\",\"span\":\"{\\\"line_start\\\":1,\\\"line_stop\\\":1,\\\"col_start\\\":6,\\\"col_stop\\\":7,\\\"path\\\":\\\"array_access.leo\\\",\\\"content\\\":\\\"x[x][y][z]\\\"}\"}"
          span:
            line_start: 1
            line_stop: 1
            col_start: 1
            col_stop: 8
            path: array_access.leo
            content: "x[x][y][z]"
      index:
        Identifier: "{\"name\":\"z\",\"span\":\"{\\\"line_start\\\":1,\\\"line_stop\\\":1,\\\"col_start\\\":9,\\\"col_stop\\\":10,\\\"path\\\":\\\"array_access.leo\\\",\\\"content\\\":\\\"x[x][y][z]\\\"}\"}"
      span:
        line_start: 1
        line_stop: 1
        col_start: 1
        col_stop: 11
        path: array_access.leo
        content: "x[x][y][z]"
  - Call:
      function:
        ArrayAccess:
          array:
            Identifier: "{\"name\":\"x\",\"span\":\"{\\\"line_start\\\":1,\\\"line_stop\\\":1,\\\"col_start\\\":1,\\\"col_stop\\\":2,\\\"path\\\":\\\"array_access.leo\\\",\\\"content\\\":\\\"x[0]()\\\"}\"}"
          index:
            Value:
              Implicit:
                - "0"
                - line_start: 1
                  line_stop: 1
                  col_start: 3
                  col_stop: 4
                  path: array_access.leo
                  content: "x[0]()"
          span:
            line_start: 1
            line_stop: 1
            col_start: 1
            col_stop: 5
            path: array_access.leo
            content: "x[0]()"
      arguments: []
      span:
        line_start: 1
        line_stop: 1
        col_start: 1
        col_stop: 7
        path: array_access.leo
        content: "x[0]()"
  - ArrayAccess:
      array:
        Call:
          function:
            Identifier: "{\"name\":\"x\",\"span\":\"{\\\"line_start\\\":1,\\\"line_stop\\\":1,\\\"col_start\\\":1,\\\"col_stop\\\":2,\\\"path\\\":\\\"array_access.leo\\\",\\\"content\\\":\\\"x()[0]\\\"}\"}"
          arguments: []
          span:
            line_start: 1
            line_stop: 1
            col_start: 1
            col_stop: 4
            path: array_access.leo
            content: "x()[0]"
      index:
        Value:
          Implicit:
            - "0"
            - line_start: 1
              line_stop: 1
              col_start: 5
              col_stop: 6
              path: array_access.leo
              content: "x()[0]"
      span:
        line_start: 1
        line_stop: 1
        col_start: 1
        col_stop: 7
        path: array_access.leo
        content: "x()[0]"
  - Call:
      function:
        CircuitStaticFunctionAccess:
          circuit:
            Call:
              function:
                Identifier: "{\"name\":\"x\",\"span\":\"{\\\"line_start\\\":1,\\\"line_stop\\\":1,\\\"col_start\\\":1,\\\"col_stop\\\":2,\\\"path\\\":\\\"array_access.leo\\\",\\\"content\\\":\\\"x(y)::y(x)\\\"}\"}"
              arguments:
                - Identifier: "{\"name\":\"y\",\"span\":\"{\\\"line_start\\\":1,\\\"line_stop\\\":1,\\\"col_start\\\":3,\\\"col_stop\\\":4,\\\"path\\\":\\\"array_access.leo\\\",\\\"content\\\":\\\"x(y)::y(x)\\\"}\"}"
              span:
                line_start: 1
                line_stop: 1
                col_start: 1
                col_stop: 5
                path: array_access.leo
                content: "x(y)::y(x)"
          name: "{\"name\":\"y\",\"span\":\"{\\\"line_start\\\":1,\\\"line_stop\\\":1,\\\"col_start\\\":7,\\\"col_stop\\\":8,\\\"path\\\":\\\"array_access.leo\\\",\\\"content\\\":\\\"x(y)::y(x)\\\"}\"}"
          span:
            line_start: 1
            line_stop: 1
            col_start: 1
            col_stop: 8
            path: array_access.leo
            content: "x(y)::y(x)"
      arguments:
        - Identifier: "{\"name\":\"x\",\"span\":\"{\\\"line_start\\\":1,\\\"line_stop\\\":1,\\\"col_start\\\":9,\\\"col_stop\\\":10,\\\"path\\\":\\\"array_access.leo\\\",\\\"content\\\":\\\"x(y)::y(x)\\\"}\"}"
      span:
        line_start: 1
        line_stop: 1
        col_start: 1
        col_stop: 11
        path: array_access.leo
        content: "x(y)::y(x)"
  - ArrayAccess:
      array:
        TupleAccess:
          tuple:
            ArrayAccess:
              array:
                Identifier: "{\"name\":\"x\",\"span\":\"{\\\"line_start\\\":1,\\\"line_stop\\\":1,\\\"col_start\\\":1,\\\"col_stop\\\":2,\\\"path\\\":\\\"array_access.leo\\\",\\\"content\\\":\\\"x[x].0[x]\\\"}\"}"
              index:
                Identifier: "{\"name\":\"x\",\"span\":\"{\\\"line_start\\\":1,\\\"line_stop\\\":1,\\\"col_start\\\":3,\\\"col_stop\\\":4,\\\"path\\\":\\\"array_access.leo\\\",\\\"content\\\":\\\"x[x].0[x]\\\"}\"}"
              span:
                line_start: 1
                line_stop: 1
                col_start: 1
                col_stop: 5
                path: array_access.leo
                content: "x[x].0[x]"
          index:
            value: "0"
          span:
            line_start: 1
            line_stop: 1
            col_start: 1
            col_stop: 7
            path: array_access.leo
            content: "x[x].0[x]"
      index:
        Identifier: "{\"name\":\"x\",\"span\":\"{\\\"line_start\\\":1,\\\"line_stop\\\":1,\\\"col_start\\\":8,\\\"col_stop\\\":9,\\\"path\\\":\\\"array_access.leo\\\",\\\"content\\\":\\\"x[x].0[x]\\\"}\"}"
      span:
        line_start: 1
        line_stop: 1
        col_start: 1
        col_stop: 10
        path: array_access.leo
        content: "x[x].0[x]"
